### ISSUE TITLE
Add dataloader for both audio windows and embeddings

### ISSUE THREAD

**Issue by acmiyaguchi** [dsgt-birdclef/birdclef-2023#53]
We should try this out using 3 second intervals with a 1 second rolling window over the span of {10, 15, 20, 30} seconds. We have removed dependencies on tensorflow for the BirdNET model via ONNX, so we can build a pytorch data loader that reads from ogg/mp3 files and generates sequences of embeddings (e.g. batch size of n with 10 frames gives us  tensors of nx10x320; the output should be the next few frames). We should take a handful of species, and preprocess the audio so all tracks are the same length (i.e. one minute, only keep the part of the track that has the highest probability of not being noise). 

We should split this up into two pieces, a AudioPCMDataSet and a BirdNETDataSet. These should generate a sequence of tensors/numpy arrays given input data. 

Here's one example of a dataset that reads data in a linear fashion: https://github.com/dsgt-birdclef/birdclef-2022/blob/main/birdclef/models/classifier_nn/datasets.py#L196-L235

It may make the most sense to have the BirdNETDataSet derive from the AudioPCMDataset, by iterating through the results and then passing it through the ONNX model.

**Comment by acmiyaguchi** [comment]
@zemm16 Let me know if there's anything you would like clarity on -- if you're interested I can also skeleton out the classes and tests that I think make the most sense before you jump on it.

**Comment by zemm16** [comment]
@acmiyaguchi  only reason is midterm week studying is taking more of my time than I was planning. Hoping to work on this tomorrow but there's a chance I can't get to working on the project till Sunday. If you want to get this jump started feel free to create a skeleton. Otherwise, I'd like to take a stab at creating everything on my own. Two questions:

1. When you say a handful of species, do you mean to start out with?
2. What does AudioPCMDataSet refer to?

**Comment by acmiyaguchi** [comment]
I added a basic skeleton for the dataset here; there were a few changes that I made while sketching things out. I added you as a reviewer, so take a look over it and you can merge it in before you start your work.

To answer your questions:

1. By this, I mean to download a couple folders from the training dataset. The input of the dataset will be the training audio, where the dataset iterates over the files, loads the PCM data via librosa (or torch.audio), and chunks it into sequential parts that we can feed into a sequence model.
1. PCM stands for [pulse-code modulation](https://en.wikipedia.org/wiki/Pulse-code_modulation) and is a digital representation of analog audio data. This is a raw format, usually with the file extension WAV. Each sample represents an intensity of a sound wave; multiple samples can be analyzed to obtain frequency/spectral information.

Some other things to note:

- https://librosa.org is the main library to take a look at while loading the audio; it's well optimized for basic audio processing stuff
- [torch.audio](https://pytorch.org/audio/stable/index.html) looks interesting, maybe read over some of this and see if it makes sense to use. The main requirement is doing some sort of sliding window thing in the dataset. 
- You might want to build a pytorch lightning data module as part of your development and testing -- that's the place where the dataset and transforms come together
- There's a little bit of thought that needs to be put into how we'll batch the audio, and how to include meta-information like the track id, the prediction label for evaluation, and channel number (in the case of sound separated data).

Treat the PR that I created as a sketch -- the main focus is setting up a few files to work in and to introduce the main testing pattern.

**Comment by zemm16** [comment]
At the moment trying to wrap my head around audio processing and the Iterable Dataset. For the audio processing being in the span of {10,15, 20, 30} seconds, do you mean having 3 seconds splits the audio with 1 sec overlap with the next 3 seconds and then have a training example of 10 seconds, and then do this process also for 15,20,and 30 seconds? So when you say keep 1 min of the track, would this multiple training examples if we are doing this process for 15,20,30 seconds? In addition, it seems the files have a huge variance of length. First one I opened is only 11 seconds.

**Comment by acmiyaguchi** [comment]
https://github.com/dsgt-birdclef/birdclef-2023/issues/31
